File: maximum-subarray/maximum-subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        
        max_so_far = -float("inf")
        max_till_here = 0
        
        for i in range(len(nums)):
            max_till_here += nums[i]
            
            if(max_till_here < 0):
                max_till_here = 0
            
            if(max_so_far < max_till_here):
                max_so_far = max_till_here
            
        if(max_so_far == 0):
            return max(nums)
        
        return max_so_far
            

File: maximum-subarray/maximum-subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        for(int i = 1; i < nums.size(); i++) {
            int nx = nums[i] + nums[i-1];
            
            if(nx > 0) {
                nums[i] = max(nums[i], nx);
            }
        }
        
        int res = -1*INT_MAX;
        for(int i = 0; i < nums.size(); i++) {
            res = max(res, nums[i]);
        }
        
        return res;
        
    }
};